File: crates/lmntalc/assets/lib/java/lmntal.java
import java.util.*;
import java.util.stream.Stream;

interface Rule {
    boolean apply();
}

enum DataType {
    PLAIN,
    INT,
    FLOAT,
    CHAR,
    STRING,
}

abstract class Atom {
    protected int name;

    public int getName() {
        return name;
    }

    public void setName(int name) {
        this.name = name;
    }

    public abstract int getArity();

    public Link at(int index) {
        throw new UnsupportedOperationException();
    }

    public boolean isInt() {
        return false;
    }

    public long getInt() {
        throw new UnsupportedOperationException();
    }

    public void setInt(long value) {
        throw new UnsupportedOperationException();
    }

    public boolean isFloat() {
        return false;
    }

    public double getFloat() {
        throw new UnsupportedOperationException();
    }

    public void setFloat(double value) {
        throw new UnsupportedOperationException();
    }

    public void setChar(char value) {
        throw new UnsupportedOperationException();
    }

    public static final class Link {
        Atom atom;
        int pos;

        public boolean isInt() {
            return atom.isInt();
        }

        public long getInt() {
            return atom.getInt();
        }

        public boolean isFloat() {
            return atom.isFloat();
        }

        public double getFloat() {
            return atom.getFloat();
        }
    }
}

final class Hyperlink extends Atom {
    private final HashSet<NormalAtom> atoms;

    public Hyperlink(int name) {
        this.atoms = new HashSet<>();
        this.name = name;
    }

    public void add(NormalAtom atom, int pos) {
        atoms.add(atom);
        atom.set(pos, this);
    }

    public void remove(NormalAtom atom, int pos) {
        atoms.remove(atom);
        atom.set(pos, null);
        if (atoms.isEmpty()) {
            AtomStore.INSTANCE.removeHyperlink(this);
        }
    }

    @Override
    public int getArity() {
        return atoms.size();
    }
}

final class NormalAtom extends Atom {
    private final Link[] args;
    private Object data;
    private DataType type;
    private boolean isRemoved;

    public NormalAtom(int name, int arity) {
        this.name = name;
        this.type = DataType.PLAIN;
        this.args = new Link[arity];
        for (int i = 0; i < arity; i++) {
            args[i] = new Link();
        }
    }

    public static boolean equals(NormalAtom... atoms) {
        for (int i = 0; i < atoms.length - 1; i++) {
            if (atoms[i] != atoms[i + 1]) {
                return false;
            }
        }
        return true;
    }

    public boolean isRemoved() {
        return isRemoved;
    }

    public void remove() {
        isRemoved = true;
    }

    public void restore() {
        isRemoved = false;
    }

    public NormalAtom getAtomAtPort(int port, int name, int arity) {
        var atom = at(port);
        if (atom == null) {
            return null;
        }
        if (atom.atom.getName() == name && atom.atom.getArity() == arity) {
            return (NormalAtom) atom.atom;
        }
        return null;
    }

    public Link at(int index) {
        return args[index];
    }

    public void set(int index, Hyperlink atom) {
        args[index].atom = atom;
        args[index].pos = -1;
    }

    public void set(int index, NormalAtom atom, int pos) {
        args[index].atom = atom;
        args[index].pos = pos;
    }

    public long getInt() {
        return (long) data;
    }

    public double getFloat() {
        return (double) data;
    }

    public char getChar() {
        return (char) data;
    }

    public String getString() {
        return (String) data;
    }

    public boolean isPlain() {
        return type == DataType.PLAIN;
    }

    public boolean isInt() {
        return type == DataType.INT;
    }

    public void setInt(long value) {
        this.data = value;
        this.type = DataType.INT;
    }

    public boolean isFloat() {
        return type == DataType.FLOAT;
    }

    public void setFloat(double value) {
        this.data = value;
        this.type = DataType.FLOAT;
    }

    public boolean isChar() {
        return type == DataType.CHAR;
    }

    public void setChar(char value) {
        this.data = value;
        this.type = DataType.CHAR;
    }

    public boolean isString() {
        return type == DataType.STRING;
    }

    public void setString(String value) {
        this.data = value;
        this.type = DataType.STRING;
    }

    public int getArity() {
        return args.length;
    }

    public void removeAt(int index) {
        AtomStore.INSTANCE.removeAtom((NormalAtom) args[index].atom);
        args[index].atom = null;
    }

    public void cloneFrom(NormalAtom atom) {
        this.data = atom.data;
        this.type = atom.type;
        System.arraycopy(atom.args, 0, this.args, 0, args.length);
    }

    @Override
    public String toString() {
        return switch (type) {
            case INT -> Long.toString(getInt());
            case FLOAT -> Double.toString(getFloat());
            default -> Main.getName(name);
        };
    }

    enum DataType {
        PLAIN,
        INT,
        FLOAT,
        CHAR,
        STRING,
    }
}

class AtomStore {
    public static AtomStore INSTANCE = new AtomStore();
    private final HashMap<Integer, List<NormalAtom>> atoms;
    private final HashMap<Integer, Queue<NormalAtom>> reusableAtoms;
    private final HashSet<Hyperlink> hyperlinks;
    private final Queue<Hyperlink> reusableHyperlinks;

    public AtomStore() {
        this.atoms = new HashMap<>();
        this.hyperlinks = new HashSet<>();
        this.reusableAtoms = new HashMap<>();
        this.reusableHyperlinks = new ArrayDeque<>();
    }

    public NormalAtom createAtom(int name, int arity) {
        if (reusableAtoms.containsKey(arity)) {
            var queue = reusableAtoms.get(arity);
            if (!queue.isEmpty()) {
                var atom = queue.poll();
                atom.setName(name);
                atom.restore();
                return atom;
            }
        }
        var atom = new NormalAtom(name, arity);
        if (!atoms.containsKey(arity)) {
            atoms.put(arity, new ArrayList<>());
        }
        atoms.get(arity).add(atom);
        return atom;
    }

    public Hyperlink createHyperlink(int name) {
        if (!reusableHyperlinks.isEmpty()) {
            var hl = reusableHyperlinks.poll();
            hl.setName(name);
            return hl;
        }
        var hl = new Hyperlink(name);
        hyperlinks.add(hl);
        return hl;
    }

    public void removeHyperlink(Hyperlink hl) {
        reusableHyperlinks.add(hl);
    }

    public void removeAtom(NormalAtom atom) {
        if (!reusableAtoms.containsKey(atom.getArity())) {
            reusableAtoms.put(atom.getArity(), new ArrayDeque<>());
        }
        atom.remove();
        var queue = reusableAtoms.get(atom.getArity());
        queue.add(atom);
    }

    public Stream<NormalAtom> findAtom(int name, int arity) {
        if (!atoms.containsKey(arity)) {
            return Stream.empty();
        }
        return atoms.get(arity)
                .stream()
                .filter(atom -> atom.getName() == name && !atom.isRemoved());
    }

    public String dumpAtoms() {
        var sb = new StringBuilder();
        var visited = new HashSet<NormalAtom>();
        for (var atoms_by_arity : atoms.values()) {
            for (var atom : atoms_by_arity) {
                if (!atom.isPlain() || atom.isRemoved() || visited.contains(atom)) {
                    continue;
                }
                dfsDump(atom, sb, visited);
            }
        }
        return sb.toString();
    }

    private void dfsDump(NormalAtom cur, StringBuilder sb, HashSet<NormalAtom> visited) {
        visited.add(cur);
        sb.append(cur.toString());
        sb.append("(");
        var count = 0;
        for (int i = 0; i < cur.getArity(); i++) {
            var atom = cur.at(i).atom;
            if (atom instanceof Hyperlink) {
                sb.append(atom.name);
                sb.append(",");
                count++;
            } else if (atom instanceof NormalAtom nr) {
                if (!visited.contains(nr)) {
                    dfsDump(nr, sb, visited);
                    sb.append(",");
                    count++;
                }
            }
        }
        sb.deleteCharAt(sb.length() - 1);
        if (count != 0) {
            sb.append(")");
        }
    }

    public NormalAtom cloneAtom(NormalAtom atom, int port) {
        var newAtom = createAtom(atom.at(port).atom.getName(), atom.at(port).atom.getArity());
        newAtom.cloneFrom((NormalAtom) atom.at(port).atom);
        return newAtom;
    }
}

File: crates/lmntalc/assets/lib/python/lmntal.py
from __future__ import annotations
import random
import copy
from typing import (
    Any,
    Generator,
    List,
)


class Atom(object):
    name: int
    pass


class Link:
    atom: Atom
    port: int

    def __init__(self, atom: Atom, port: int):
        self.atom = atom
        self.port = port

    def is_int(self):
        if isinstance(self.atom, NormalAtom):
            return self.atom.is_int()
        return False

    def get_int(self) -> int:
        if isinstance(self.atom, NormalAtom):
            return self.atom.get_int()
        return 0


class NormalAtom(Atom):
    arity: int
    data: Any
    args: List[Link]
    removed: bool = False

    def __init__(self, name: int, arity: int):
        self.name = name
        self.arity = arity
        self.data = None
        self.args = list()
        for i in range(arity):
            self.args.append(Link(None, 0))
        self.removed = False

    def __str__(self):
        if isinstance(self.data, int) or isinstance(self.data, float):
            return f"{self.data}"
        return name_map[self.name]

    def arity(self) -> int:
        return self.arity

    def at(self, index: int) -> Link:
        return self.args[index]

    def remove_at(self, index: int):
        self.args[index].atom.removed = True
        self.args[index].atom = None

    def is_plain(self) -> bool:
        return self.data is None

    def is_int(self) -> bool:
        return self.data is not None and isinstance(self.data, int)

    def set_int(self, data: int):
        self.data = data

    def get_int(self) -> int:
        return self.data

    def is_float(self) -> bool:
        return self.data is not None and isinstance(self.data, float)

    def is_str(self) -> bool:
        return self.data is not None and isinstance(self.data, str)


class Hyperlink(Atom):
    args: set[NormalAtom]

    def __init__(self, name: int):
        self.name = name
        self.args = set()

    def add(self, atom: NormalAtom, index: int):
        self.args.add(atom)
        atom.args[index].atom = self

    def remove(self, atom: NormalAtom, index: int):
        self.args.remove(atom)
        atom.args[index].atom = None

    def arity(self) -> int:
        return len(self.args)


class AtomStore:
    store: dict[int, list[NormalAtom]]
    hyperlinks: list[Hyperlink]

    def __init__(self):
        self.store = {}
        self.hyperlinks = []

    def create(self, name: int, arity: int) -> NormalAtom:
        if arity not in self.store:
            self.store[arity] = list()
        for atom in self.store[arity]:
            # reuse removed atom
            if atom.removed:
                atom.removed = False
                atom.name = name
                return atom
        atom = NormalAtom(name, arity)
        self.store[arity].append(atom)
        return atom

    def clone(self, atom: NormalAtom, port: int) -> NormalAtom:
        target = atom.args[port].atom # type: NormalAtom
        atom2 = self.create(target.name, target.arity)
        atom2.data = copy.deepcopy(target.data)
        for i in range(atom2.arity):
            atom2.args[i] = copy.deepcopy(target.args[i])
        return atom2

    def create_hyperlink(self, name: int) -> Hyperlink:
        hl = Hyperlink(name)
        self.hyperlinks.append(hl)
        return hl

    def remove(self, atom: NormalAtom):
        atom.removed = True

    def find(self, name: int, arity: int) -> Generator[NormalAtom, None, None]:
        if arity not in self.store:
            return
        for atom in self.store[arity]:
            if atom.removed:
                continue
            if atom.name == name and atom.arity == arity:
                yield atom

    def dump(self) -> str:
        visited = set()
        res = ""
        for _, atoms in self.store.items():
            for atom in atoms:
                if not atom.is_plain or atom.removed or (atom in visited):
                    continue
                res += self.__dfs_dump(atom, visited, res)
        return res

    def __dfs_dump(self, atom: NormalAtom, visited: set[NormalAtom], string: str) -> str:
        visited.add(atom)
        string += str(atom)
        string += "("
        count = 0
        for i in range(atom.arity):
            atom2 = atom.at(i).atom
            if isinstance(atom2, Hyperlink):
                string += name_map[atom2.name]
                string += ","
                count += 1
            elif isinstance(atom2, NormalAtom) and (atom2 not in visited):
                string = self.__dfs_dump(atom2, visited, string)
                string += ","
                count += 1
        string = string[:-1]
        if count != 0:
            string += ")"
        return string


atom_list: AtomStore = AtomStore()


def create_atom(name: int, arity: int) -> NormalAtom:
    return atom_list.create(name, arity)


def clone_atom(atom: NormalAtom, port: int) -> NormalAtom:
    return atom_list.clone(atom, port)


def create_hyperlink(name: int) -> Hyperlink:
    return atom_list.create_hyperlink(name)


def remove_atom(atom: NormalAtom):
    atom_list.remove(atom)


def find_atom(name: int, arity: int) -> Generator[NormalAtom, None, None]:
    return atom_list.find(name, arity)


def link(atom: NormalAtom, index: int, atom2: NormalAtom, index2: int):
    atom.args[index].atom = atom2
    atom.args[index].port = index2
    atom2.args[index2].atom = atom
    atom2.args[index2].port = index


def relink(atom: NormalAtom, index: int, atom2: NormalAtom, index2: int):
    atom3 = atom.args[index]
    link(atom3.atom, atom3.port, atom2, index2)


def unify(atom: NormalAtom, index: int, atom2: NormalAtom, index2: int):
    atom3 = atom.args[index]
    atom4 = atom2.args[index2]
    link(atom3.atom, atom3.port, atom4.atom, atom4.port)


def get_atom_at_port(atom: NormalAtom, index: int, name: str, arity: int) -> NormalAtom | None:
    atom2 = atom.at(index)
    if isinstance(atom2, NormalAtom) and atom2.name == name and atom2.arity == arity:
        return atom2
    return None


def get_hyperlink_at_port(atom: NormalAtom, index: int) -> Hyperlink | None:
    atom2 = atom.at(index)
    if isinstance(atom2, Hyperlink):
        return atom2
    return None


def equals(*atoms: Atom) -> bool:
    for i in range(len(atoms) - 1):
        if atoms[i] != atoms[i + 1]:
            return False
    return True


def dump_atoms():
    print(atom_list.dump())

